dev-tools/xtask/src/a4x2_package.rs

```diff
@@ -0,0 +1,382 @@
+        };
+
+        let src_dir = Utf8PathBuf::try_from(env::current_dir()?)?;
+
+        // Delete any results from previous runs. We don't mind if there's errors.
+        fs::remove_dir_all(&work_dir).ok();
+        // Create work dir
+        fs::create_dir_all(&work_dir)?;
+        sh.change_dir(&work_dir);
+
+        // Output. Maybe in CI we want this to be /out
+        let out_dir = work_dir.join("a4x2-package-out");
+        fs::create_dir_all(&out_dir)?;
+
+        // Clone of omicron that we can modify without causing trouble for anyone
+        let omicron_dir = work_dir.join("omicron");
+        fs::create_dir_all(&omicron_dir)?;
+
+        Environment {
+            cargo,
+            home_dir,
+            work_dir,
+            src_dir,
+            out_dir,
+            omicron_dir,
+            in_ci,
+        }
+    };
+
+    prepare_source(&sh, &env)?;
+
+    if args.end_to_end_tests {
+        build_end_to_end_tests(&sh, &env)?;
+    }
+
+    if args.live_tests {
+        build_live_tests(&sh, &env)?;
+    }
+
+    // This needs to happen last because it messes with the working tree in a
+    // way that end-to-end tests doesnt like when building
+    build_a4x2(&sh, &env)?;
+
+    create_output_artifact(&sh, &env)?;
+
+    Ok(())
+}
+
+/// Clone local working tree into work dir. This is nabbed from
+/// buildomat-at-home
+fn prepare_source(sh: &Shell, env: &Environment) -> Result<()> {
+    cmd!(sh, "banner 'prepare'").run()?;
+
+    // Get a ref we can checkout from the local working tree
+    let treeish = {
+        let _popdir = sh.push_dir(&env.src_dir);
+        let mut treeish = cmd!(sh, "git stash create").read()?;
+        if treeish.is_empty() {
+            // nothing to stash
+            eprintln!("Nothing to stash, using most recent commit for clone");
+            treeish = cmd!(sh, "git rev-parse HEAD").read()?;
+        }
+        assert_eq!(
+            treeish.len(),
+            40,
+            "treeish should be a 40-character commit hash"
+        );
+        assert!(
+            treeish.chars().all(|c| c.is_ascii_hexdigit()),
+            "treeish should be a 40-character commit hash"
+        );
+        treeish
+    };
+
+    // Clone the local source tree into the packaging work dir
+    let _popdir = sh.push_dir(&env.omicron_dir);
+    cmd!(sh, "git init").run()?;
+
+    let src_dir = &env.src_dir;
+    cmd!(sh, "git remote add origin {src_dir}").run()?;
+
+    cmd!(sh, "git fetch origin {treeish}").run()?;
+    cmd!(sh, "git checkout {treeish}").run()?;
+
+    // XXX is this needed in CI?
+    cmd!(sh, "./tools/install_builder_prerequisites.sh -yp").run()?;
+
+    Ok(())
+}
+
+fn build_end_to_end_tests(sh: &Shell, env: &Environment) -> Result<()> {
+    // I'm not confident this does what it should, because I don't know much
+    // about the end to end tests. This is just going from what the shell script
+    // did.
+    cmd!(sh, "banner 'end to end'").run()?;
```

Comment:
hmm, I know `banner` has a character limit, might be worth checking whether this command works.

Reply:
it does, i ran into that char limit earlier in dev but this one is fine